### PR TITLE
Adds a new completion state to the deployer

### PR DESCRIPTION
**Pull request by finefuture**
## What is the purpose of the change
Publish started event after all action done.

before:
image
image

after:
image

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by finefuture** [comment]
Adjust dubbo-samples in pr:  https://github.com/apache/dubbo-samples/pull/1015

**Comment by finefuture** [comment]
> Change the event order might not a good idea in 3.2.x. Some users have depended on this order to export or refer services.

It might be better to add some other events to implement this functionality for the user.For example, ModuleReadyEvent.

**Comment by AlbumenJ** [comment]
> > Change the event order might not a good idea in 3.2.x. Some users have depended on this order to export or refer services.
> 
> It might be better to add some other events to implement this functionality for the user.For example, ModuleReadyEvent.

Yes, and we can make it in 3.3.x.

**Comment by AlbumenJ** [comment]
> > Change the event order might not a good idea in 3.2.x. Some users have depended on this order to export or refer services.
> 
> It might be better to add some other events to implement this functionality for the user.For example, ModuleReadyEvent.

I think add some newly events would be a better option.

**Comment by finefuture** [comment]
> > > Change the event order might not a good idea in 3.2.x. Some users have depended on this order to export or refer services.
> > 
> > 
> > It might be better to add some other events to implement this functionality for the user.For example, ModuleReadyEvent.
> 
> I think add some newly events would be a better option.

How about adding a new state, like completion.

**Comment by AlbumenJ** [comment]
> > > > Change the event order might not a good idea in 3.2.x. Some users have depended on this order to export or refer services.
> > > 
> > > 
> > > It might be better to add some other events to implement this functionality for the user.For example, ModuleReadyEvent.
> > 
> > 
> > I think add some newly events would be a better option.
> 
> How about adding a new state, like completion.

ok

File: dubbo-common/src/main/java/org/apache/dubbo/common/deploy/DeployListener.java [hunk 1]
```diff
@@ -28,6 +28,8 @@ public interface DeployListener<E extends ScopeModel> {
 
     void onStarted(E scopeModel);
 
+    void onCompletion(E scopeModel);
+
     void onStopping(E scopeModel);
 
     void onStopped(E scopeModel);
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/deploy/DeployListener.java:35]: I think we should add some comment on these statuses.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/deploy/DeployListener.java (hunk 1 above)
Comment (finefuture) [reply dubbo-common/src/main/java/org/apache/dubbo/common/deploy/DeployListener.java:35]: Okay, I‘ll add some comment.

**Pull request merged**